prow/entrypoint/run_test.go

```diff
@@ -127,6 +130,44 @@ func TestOptions_Run(t *testing.T) {
 			expectedMarker: "0",
 			expectedCode:   0,
 		},
+
+		{
+			name:      "interrupt, propagate child error",
```

Comment:
If your shell script just exits, you won't need to interrupt it, right?

Reply:
Are those important? Why not just write a script that exits with your code without sleeping or traps?

Reply:
well, but I want to test out when the entrypoint receives a SIGTERM it exits with the child error, not the 130. As it is now, a SIGTERM to the entrypoint exits the entrypoint with the 130 error, that's what I want to avoid with this PR

Reply:
Sorry, I was probably not precise enough with the description of the PR.

Reply:
I see. I'd argue the invariant we want to test with this new functionality is "if `entrypoint` is configured to do so, it exits with the same exit code as the child." It's not important how the child came to exit with that code, just that `entrypoint` faithfully returns whatever the child did.

With that being said, I can understand that this is a useful integration test, but I think it is not necessary to test implementation specifics of `SIGTERM` and `trap` to be confident that your new option functions as it should.

In general, dependency injection points that exist in production code and take only half of the possible code paths in normal cases, as opposed to a different set of code paths in testing, are to my opinion fragile. The `Options` as they exist today are meant to be static flag-driven options passed to the binary, and the assumptions we hold about them are fairly static. Having to add in a dynamic value that's not actually an option to end users smells like the wrong level of abstraction here.

The test calls `Run()` - perhaps we can factor the code such that some new method `options.internalRun()` takes a context as a parameter, which we'd be able to mess with in testing, and the user-facing interface here does not change.

---

One other thing - let's think about how we might improve `Options.Validate()` for the new option - it seems like an error to specify `o.AlwaysZero` as well as `o.PropagateErrorCode`, and perhaps also an error if there are other previous markers that we'd otherwise forward?

Reply:
I think the current version of this PR looks ready to merge! When you're ready, go ahead and remove the "draft" marking and I will LGTM :)

Reply:
This test case is flaky: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/30428/pull-test-infra-unit-test/1697310401185714176

@Prucek is there a way to make this test case less flaky?

Reply:
It really seems that the test is flaky:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/30574/pull-test-infra-unit-test/1697027723278422016
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-test-infra-unit-test/1697509322486976512
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-test-infra-unit-test/1696934500996485120

I've never investigated a flaky test, so I don't really know what to do, @stevekuznetsov could you maybe help me with this one? I think the problem is that sometimes the wrapped command has not yet started and it already received the interrupt, I assume if we increase the time when the interrupt is sent in https://github.com/kubernetes/test-infra/blob/master/prow/entrypoint/run_test.go#L222 to maybe 500 ms it could help. What do you think? Thank you for your help!

Reply:
#30655 Related PR

Reply:
Sorry, I just saw this - left a comment on the PR, thanks for looking into it @Prucek